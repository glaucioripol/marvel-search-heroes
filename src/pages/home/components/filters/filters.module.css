.filters {
  display: flex;
  justify-content: space-between;
  margin-block: 48px;
}

.filters .filters__total_results {
  color: var(--grey-4);
  font-size: var(--font-x-sm);
}

.filters .filters__actions {
  align-items: center;
  display: flex;
  gap: 32px;
}
