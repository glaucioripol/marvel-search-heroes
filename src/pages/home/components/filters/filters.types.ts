export type FiltersProperties = Readonly<{
  totalItems: number;
  showJustFavorites: boolean;
  onPressJustFavoriteButton: () => void;
  onAlphabeticalOrderChange: (isChecked: boolean) => void;
  isAlphabeticalOrderEnabled: boolean;
  areThereFavoriteHeroes: boolean;
}>;
