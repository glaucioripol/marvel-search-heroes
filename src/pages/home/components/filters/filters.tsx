import { ShowJustFavoritesButton, ToggleButton } from "@/components";

import styles from "./filters.module.css";

import { FiltersProperties } from "./filters.types";

export function Filters({
  totalItems,
  showJustFavorites,
  onAlphabeticalOrderChange,
  isAlphabeticalOrderEnabled,
  areThereFavoriteHeroes,
  onPressJustFavoriteButton,
}: FiltersProperties) {
  return (
    <div className={styles.filters} data-testid="filters-wrapper">
      <div>
        <span className={styles.filters__total_results}>
          Encontrados {totalItems} heróis
        </span>
      </div>

      <div className={styles.filters__actions}>
        <div className="text_with_icon">
          <img
            src="/assets/ic_heroi.svg"
            alt="Desenho que representa um herói, uma pessoa de capa, com a mão na cintura e todo o desenho é em vermelho."
            title="Desenho que representa um herói, uma pessoa de capa, com a mão na cintura e todo o desenho é em vermelho."
            width={18}
            height={27}
          />

          <span>Ordenar por nome - A/Z</span>
        </div>

        <div>
          <ToggleButton
            isChecked={isAlphabeticalOrderEnabled}
            onChange={onAlphabeticalOrderChange}
          />
        </div>

        <div>
          <ShowJustFavoritesButton
            isActive={showJustFavorites}
            onClick={onPressJustFavoriteButton}
            disabled={!areThereFavoriteHeroes}
          />
        </div>
      </div>
    </div>
  );
}
