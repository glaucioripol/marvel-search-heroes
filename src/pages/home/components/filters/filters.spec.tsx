import {
  defaultRender,
  fireEvent,
  vi,
  describe,
  expect,
  it,
} from "@/__tests__";

import { Filters } from "./filters";
import { FiltersProperties } from "./filters.types";

const defaultProperties = () =>
  ({
    totalItems: 10,
    showJustFavorites: false,
    onAlphabeticalOrderChange: vi.fn(),
    isAlphabeticalOrderEnabled: false,
    areThereFavoriteHeroes: false,
    onPressJustFavoriteButton: vi.fn(),
  }) as FiltersProperties;

describe("Filters component", () => {
  const sutFilters = (customProperties = {} as Partial<FiltersProperties>) => {
    const currentDefaultProperties = defaultProperties();
    const props = { ...currentDefaultProperties, ...customProperties };
    const renderedComponent = defaultRender(<Filters {...props} />);

    return {
      ...renderedComponent,
      onAlphabeticalOrderChangeMock:
        props.onAlphabeticalOrderChange ??
        currentDefaultProperties.onAlphabeticalOrderChange,
      onPressJustFavoriteButtonMock:
        props.onPressJustFavoriteButton ??
        currentDefaultProperties.onAlphabeticalOrderChange,
    };
  };

  it("should render total items correctly", () => {
    const { getByText } = sutFilters();
    const totalItemsElement = getByText(
      `Encontrados ${defaultProperties().totalItems} herÃ³is`,
    );
    expect(totalItemsElement).toBeInTheDocument();
  });

  it("should call onAlphabeticalOrderChange when toggle button is clicked", () => {
    const { getByRole, onAlphabeticalOrderChangeMock } = sutFilters();
    const toggleButton = getByRole("checkbox");
    fireEvent.click(toggleButton);
    expect(onAlphabeticalOrderChangeMock).toHaveBeenCalled();
  });

  it("should call onPressJustFavoriteButton when show just favorites button is clicked", () => {
    const { getByTestId, onPressJustFavoriteButtonMock } = sutFilters({
      areThereFavoriteHeroes: true,
    });
    const showJustFavoritesButton = getByTestId("just-liked-button");
    fireEvent.click(showJustFavoritesButton);
    expect(onPressJustFavoriteButtonMock).toHaveBeenCalled();
  });

  it("should match snapshot", () => {
    const { asFragment } = sutFilters();
    expect(asFragment()).toMatchSnapshot();
  });
});
