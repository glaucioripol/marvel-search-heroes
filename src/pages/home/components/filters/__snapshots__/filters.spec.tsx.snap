// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filters component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="_filters_a2f514"
    data-testid="filters-wrapper"
  >
    <div>
      <span
        class="_filters__total_results_a2f514"
      >
        Encontrados 10 heróis
      </span>
    </div>
    <div
      class="_filters__actions_a2f514"
    >
      <div
        class="text_with_icon"
      >
        <img
          alt="Desenho que representa um herói, uma pessoa de capa, com a mão na cintura e todo o desenho é em vermelho."
          height="27"
          src="/assets/ic_heroi.svg"
          title="Desenho que representa um herói, uma pessoa de capa, com a mão na cintura e todo o desenho é em vermelho."
          width="18"
        />
        <span>
          Ordenar por nome - A/Z
        </span>
      </div>
      <div>
        <div
          aria-checked="false"
          aria-label="Ativar"
          aria-labelledby="Ativar"
          class="_toggle_button_7fc16a "
          data-testid="toggle-button-default"
          role="checkbox"
          tabindex="0"
          title="Ativar"
        />
      </div>
      <div>
        <button
          aria-label="Exibir somente favoritos"
          class="_just_liked_button_62679c text_with_icon  _just_liked_button--disabled_62679c"
          data-testid="just-liked-button"
          disabled=""
          title="Exibir somente favoritos"
          type="button"
        >
          <img
            alt="Imagem de um coração, representando um favorito."
            height="18"
            src="/assets/favorito_01.svg"
            width="18"
          />
          <span>
            Somente favoritos
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
